client/app/pages/queries/hooks/useQueryExecute.js

```diff
@@ -1,79 +1,90 @@
-import { useState, useCallback, useMemo, useEffect, useRef } from "react";
-import { noop, includes } from "lodash";
-import useQueryResult from "@/lib/hooks/useQueryResult";
+import { useReducer, useCallback, useEffect } from "react";
 import location from "@/services/location";
 import recordEvent from "@/services/recordEvent";
+import { ExecutionStatus } from "@/services/query-result";
 
 function getMaxAge() {
   const { maxAge } = location.search;
   return maxAge !== undefined ? maxAge : -1;
 }
 
+const reducer = (prevState, updatedProperty) => ({
+  ...prevState,
+  ...updatedProperty,
+});
+
+// This is currently specific to a Query page, we can refactor
+// it slightly to make it suitable for dashboard widgets instead of the other solution it
+// has in there.
 export default function useQueryExecute(query) {
-  // Query result should be initialized only once on component mount
-  const initializeQueryResultRef = useRef(() =>
-    query.hasResult() || query.paramsRequired() ? query.getQueryResult(getMaxAge()) : null
-  );
-  const [queryResult, setQueryResult] = useState(initializeQueryResultRef.current());
-  initializeQueryResultRef.current = noop;
+  const [executionState, setExecutionState] = useReducer(reducer, {
+    queryResult: null,
+    isExecuting: false,
+    executionStatus: null,
+    isCancelling: false,
+    cancelCallback: null,
+    error: null,
+  });
+
+  const executeQuery = useCallback(
+    (maxAge = 0, queryExecutor) => {
+      let newQueryResult;
+      if (queryExecutor) {
+        newQueryResult = queryExecutor();
+      } else {
+        newQueryResult = query.getQueryResult(maxAge);
+      }
 
-  const queryResultData = useQueryResult(queryResult);
-  const isQueryExecuting = useMemo(() => !!queryResult && !includes(["done", "failed"], queryResultData.status), [
-    queryResult,
-    queryResultData.status,
-  ]);
+      setExecutionState({
+        updatedAt: newQueryResult.getUpdatedAt(),
+        isExecuting: true,
+        cancelCallback: () => {
+          recordEvent("cancel_execute", "query", query.id);
+          setExecutionState({ isCancelling: true });
+          newQueryResult.cancelExecution();
+        },
+      });
 
-  const [isExecutionCancelling, setIsExecutionCancelling] = useState(false);
+      const onStatusChange = status => {
+        setExecutionState({ updatedAt: newQueryResult.getUpdatedAt(), executionStatus: status });
+      };
 
-  const executeQuery = useCallback(() => {
-    recordEvent("execute", "query", query.id);
-    setQueryResult(query.getQueryResult(0));
-  }, [query]);
+      newQueryResult
+        .toPromise(onStatusChange)
+        .then(queryResult => {
+          // TODO: this should probably belong in the QueryEditor page.
```

Comment:
A quick note: here (and in other similar places) we need to handle a case when component was destroyed during async operation (ignore all `.then`/`.catch` and callback calls otherwise all sort of `setState`s will trigger error).

Reply:
:+1: